### PR TITLE
Translations feature w/feature toggle

### PR DESCRIPTION
**Pull request by it-vegard**

### CODE REVIEW

File: src/Forms/PrepareLetterPage.jsx [hunk 1]
```diff
@@ -7,21 +7,39 @@ import PropTypes from "prop-types";
...
+            skalSendeFlereVedlegg
+              ? TEXTS.prepareLetterPage.attachmentSectionTitleTheseAttachments
+              : TEXTS.prepareLetterPage.attachmentSectionTitleThisAttachment
+          )
+      )}`}
     >
       <Systemtittel className="margin-bottom-default">
-        {index}. Legg ved {skalSendeFlereVedlegg ? "disse vedleggene" : "dette vedlegget"}
+        {`${index}. ${translate(
+          TEXTS.prepareLetterPage.attachmentSectionTitleAttachTo
```
Comment (it-vegard) [inline src/Forms/PrepareLetterPage.jsx:31]: Repeats code from just above. Extract to a const before the return.

File: src/context/i18n/index.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+};
+
+const I18nContext = createContext({});
+
+function I18nProvider({ children, loadTranslations }) {
+  const [translations, setTranslations] = useState({});
+  const [translationsForNavForm, setTranslationsForNavForm] = useState(null);
+  const [currentTranslation, setCurrentTranslation] = useState({});
+
+  useEffect(() => {
```
Comment (it-vegard) [inline src/context/i18n/index.js:25]: May cause redudant loading of translations. Check if we can remove/move this logic.

File: src/components/NavForm.jsx [hunk 3]
```diff
@@ -201,6 +201,7 @@ class NavForm extends Component {
 
   componentDidMount = () => {
     const { options, src, url, form } = this.props;
+    console.log("Options: ", options);
```
Comment (it-vegard) [inline src/components/NavForm.jsx:204]: Remove this

File: src/context/i18n/translationsMapper.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import GlobalTranslations from "../../i18nData";
+
+const mapTranslationsToFormioI18nObject = (translations) => {
```
Comment (it-vegard) [inline src/context/i18n/translationsMapper.js:3]: Look at splitting out parts of the code in util functions

File: src/components/FormMetadataEditor.test.jsx [hunk 5]
```diff
@@ -58,35 +84,65 @@ describe("FormMetadataEditor", () => {
...
+                  formio={new Formio("http://myproject.example.org")}
+                  store={{ forms: [fakeBackend.form()] }}
+                />
+              </AppConfigProvider>
+            </UserAlerterContext.Provider>
+          </AuthContext.Provider>
+        </MemoryRouter>
+      );
+
+      //let visningsSelect = context.testRenderer.root.findByType(SkjemaVisningSelect);
```
Comment (it-vegard) [inline src/components/FormMetadataEditor.test.jsx:138]: Kan fjerne kommentarer

File: src/components/FormMetadataEditor.jsx [hunk 6]
```diff
@@ -42,6 +43,21 @@ const BasicFormMetadataEditor = ({ form, onChange, usageContext }) => {
         readOnly={usageContext === "edit"}
         onChange={(event) => onChange({ ...form, properties: { ...form.properties, tema: event.target.value } })}
       />
+      <Select
+        label="Type"
+        name="form-type"
+        id="form-type"
+        value={type}
+        // readOnly={true}
```
Comment (it-vegard) [inline src/components/FormMetadataEditor.jsx:51]: Fjern utkommentert prop

File: src/Forms/index.test.jsx [hunk 7]
```diff
@@ -90,13 +94,13 @@ describe("FormsRouter", () => {
...
     jest.runOnlyPendingTimers();
-    const formPage = await context.waitForComponent(FormsListPage);
+    await context.waitForComponent(FormsListPage);
     clickHovedknapp("Lag nytt skjema");
     expect(routeLocation().pathname).toEqual("/forms/new");
     await context.waitForComponent(NewFormPage);
   });
 
-  const wait = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
+  //const wait = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
```
Comment (it-vegard) [inline src/Forms/index.test.jsx:103]: Kan fjernes

File: src/translations/TranslationsByFormPage.jsx [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+import { flattenComponents } from "../util/forsteside";
+import FormBuilderLanguageSelector from "../context/i18n/FormBuilderLanguageSelector";
+import { Hovedknapp, Knapp } from "nav-frontend-knapper";
+import TranslationsFormPage from "./TranslationsFormPage";
+import { useTranslations } from "../context/i18n";
+import { LanguagesProvider } from "../context/languages";
+import i18nData from "../i18nData";
+import useRedirectIfNoLanguageCode from "../hooks/useRedirectIfNoLanguageCode";
+
+const getAllTextsForForm = (form) =>
```
Comment (it-vegard) [inline src/translations/TranslationsByFormPage.jsx:13]: Refactor to separate file

Comment (it-vegard) [reply src/translations/TranslationsByFormPage.jsx:13]: Also add tests

File: src/translations/TranslationsByFormPage.jsx [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        ...allTextsForForm,
+        ...Object.keys(component)
+          .filter((key) => component[key] !== undefined)
+          .reduce((textsForComponent, key) => {
+            if (key === "values") {
+              return [
+                ...textsForComponent,
+                ...component[key].map((value) => ({
+                  text: value,
+                  type: value.length < 80 ? "text" : "textarea",
```
Comment (it-vegard) [inline src/translations/TranslationsByFormPage.jsx:36]: Extract function for deciding between text or textarea to helper function

File: src/translations/TranslationsByFormPage.jsx [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+                ...textsForComponent,
+                { text: component[key], type: component[key].length < 80 ? "text" : "textarea" },
+              ];
+            }
+          }, []),
+      ];
+    }, [])
+    .filter(
+      (component, index, currentComponents) =>
+        index === currentComponents.findIndex((currentComponent) => currentComponent.text === component.text)
```
Comment (it-vegard) [inline src/translations/TranslationsByFormPage.jsx:50]: This is code for filtering out duplicate entries. Should either add comments or extract to well named helper functions

File: src/hooks/useForms.js [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+
+  const loadTranslationsForEditPage = async (formPath) => {
+    return loadTranslationsForForm(formPath)
+      .then((translations) =>
+        translations
+          .map((translation) => ({
+            ...translation.data,
+            id: translation._id,
+          }))
+          .reduce(
```
Comment (it-vegard) [inline src/hooks/useForms.js:158]: We need tests for this + refactoring

File: src/hooks/useForms.js [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+            {}
+          )
+      )
+      .then((translations) => {
+        console.log("loadTranslationsForEditPage", translations);
+        return translations;
+      });
+  };
+
+  const deleteLanguage = async (id) => {
```
Comment (it-vegard) [inline src/hooks/useForms.js:208]: Rename to deleteTranslation

File: src/customComponents/components/Checkbox.jsx [hunk 13]
```diff
@@ -28,13 +28,13 @@ const CheckboxWrapper = class extends Component {
...
-    const component = this.props.component;
+    const { component, translate } = this.props;
     return (
       <Checkbox
         checkboxRef={this.props.checkboxRef}
         aria-describedby={`${component.key}-error`}
-        label={component.label}
-        onChange={(event) => this.setValue(!!this.state.value ? null : "ja")}
+        label={translate(component.label)}
+        onChange={(event) => this.setValue(this.state.value === "on" ? "off" : "on")}
```
Comment (it-vegard) [inline src/customComponents/components/Checkbox.jsx:37]: Check if this is merged to master before, and decide which code is correct

Comment (wen7i) [reply src/customComponents/components/Checkbox.jsx:37]: It was merge conflict and our code is outdated. Fixed.

File: src/Forms/PrepareLetterPage.jsx (hunk 1 above)
Comment (wen7i) [reply src/Forms/PrepareLetterPage.jsx]: Fixed

File: src/hooks/useForms.js (hunk 12 above)
Comment (wen7i) [reply src/hooks/useForms.js:208]: Done

File: src/Forms/index.test.jsx (hunk 7 above)
Comment (wen7i) [reply src/Forms/index.test.jsx:103]: Done

File: src/components/NavForm.jsx (hunk 3 above)
Comment (wen7i) [reply src/components/NavForm.jsx:204]: Done

**Pull request merged**